### DISCUSSION TITLE
append to detached object

### DISCUSSION THREAD

**Discussion by ostapkostyk** [sqlalchemy/sqlalchemy#9074] [Usage Questions] [answered]
I try to build a correct architecture for GUI application that uses SQLAlchemy.  
A user should be able to add elements to the tree.  
On one hand, the documentation says that it is not desirable to keep objects in a detached state. On the other hand, it says that this is not a good practice to keep session open and idling for a long time. In case of GUI application these look like controversial statements and I try to deal with that.
One way is to keep user input separately from the ORM objects and copy when the user confirms changes, but I find it senseless.  
 
Then, the other important point is to separate session operations from operations with ORM objects itself.  
Also, I'd like to avoid all sort of problems with connectivity issues when keeping the session open for a long time.  

Therefore I tried an approach to load existing data to ORM objects and detach them (to keep the session closed). When a user wants to add a new object (on push button event), I create an empty ORM object and send it to a function or a method that creates a session within "with" statement, add that object to a session and commit. So that I run short living session.  
But I can't create a related ORM object in relation to a detached object (due to DetachedInstanceError exception).  
Here is a code:

```python
from sqlalchemy import create_engine, Column, Integer, String, MetaData, ForeignKey, text, inspect, exc
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column, relationship, declared_attr
from sqlalchemy.orm import sessionmaker

engine = create_engine("postgresql+psycopg://postgres:my_password@localhost:5432/MyDB", echo=True)

Session = sessionmaker(bind=engine)

class Base(DeclarativeBase):
    pass

class User(Base):
    __tablename__ = "users"

    id : Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str] = mapped_column(String)

    addresses: Mapped[list["Address"]] = relationship(back_populates="user", cascade="all, delete-orphan")

    def __repr__(self):
        return f"User: {self.id}, {self.name}"


class Address(Base):
    __tablename__ = "addresses"

    id: Mapped[int] = mapped_column(primary_key=True)
    user_fk: Mapped[int] = mapped_column(ForeignKey("users.id"))
    address: Mapped[str] = mapped_column(String)

    user: Mapped["User"] = relationship(back_populates="addresses")


def get_session():
    with Session() as session:
        yield session


def save(a):                # primitive example but can be isolated in separate module
    for session in get_session():
        session.add(a)
        session.commit()


user = User(name='user1')
save(user)  # works well

user.addresses.append(Address(address="address1"))  # this throws an exception 
```
exception:
```
sqlalchemy.orm.exc.DetachedInstanceError: Parent instance <User at 0x20537308370> is not bound to a Session; lazy load operation of attribute 'addresses' cannot proceed (Background on this error at: https://sqlalche.me/e/20/bhk3)
```

What would be a good practice to keep short-living session and isolate session-related code from operations with ORM data?  
Code snippets are very much appreciated.

**Comment by zzzeek** [comment, marked as answer]
I like the thinking you are doing here and my two thoughts are:

1. why not keep objects attached, and call session.commit() each time the user performs a GUI operation (or perhaps intermittently using a background thread)?  that would keep the actual transaction closed while the database remains available.
2. if you dont actually want the user's activities to be immediately committed until they press a "save" button, that suggests the GUI is presenting a "transaction" view itself - it means you really do have a "tranasction" in play that you want to commit at some point.    This might be what you are doing, so two options in this case are:  

a. just have a real transaction in play - even this is what you want to avoid, if the GUI app is truly trying to keep a "transaction" open by waiting for the "save" button to be pushed, maybe it's appropriate.

b. use DBAPI autocommit (see https://docs.sqlalchemy.org/en/14/orm/session_transaction.html#setting-transaction-isolation-levels-dbapi-autocommit)  and set autoflush=False.   Nothing will be "written" to the database until you call .flush() or .commit().   There won't be any database transaction in play ever.    this might be a very good mode for you to use if your app is trying to do essentially a "client side transaction" of sorts.

in all cases you still have a "connection" in play, after all the connection pool in the Engine will keep a connection around unless you configured NullPool.   But for options 1. and 2b., there wont be any "idle in transaction" state.

**Reply by ostapkostyk** [reply]
I was thinking about the second option, where GUI is presenting a "transaction" view itself. I'd like to use ORM objects in a model-view-controller pattern. So, at the beginning, the application loads data from the database to ORM objects and presents them to the user. Then, the user modifies some fields or adds new objects and finally commits everything by pressing "Save" button. I experimented with this approach a bit. Problems start if ORM objects expire unintentionally, for example, due to connectivity issues or database restart, whatever (I'm not aware of all possible cases). Then I need to rollback the session and this overwrites all user changes in these ORM objects. So the user input is lost.  Probably this is the biggest issue for me.
The other case - if I want to implement two-phase-lock mechanism for concurrent access and would like to refresh just two columns without losing the modified data in the object - seems this is also not possible.  
Keeping user input separate from ORM objects looks ugly because ORM class already defines a dataset.  
As I understand, keeping objects detached guarantees that the modified data is not lost in case of problems with the session. 
But then, of course, these objects do not reflect the state of the database and I need to do something to synchronize them or just to overwrite data in the database (maybe to use locks to prevent conflicts). From these thoughts born the idea to keep objects detached and add them to a fresh session to commit changes. But I'm not sure if this is a good approach.

**Reply by CaselIT** [reply]
in your example you would need to first re-add the user to a session before appending a new address, since the addresses relation is not loaded and the orm is strying to load it.

you can use [merge](https://docs.sqlalchemy.org/en/14/orm/session_api.html#sqlalchemy.orm.Session.merge) to transfer the changes from the detached object to the one in the session. Alternatively you can use `add` if the detached object has no changed (not 100% about the second part)

**Reply by zzzeek** [reply]
> I was thinking about the second option, where GUI is presenting a "transaction" view itself. I'd like to use ORM objects in a model-view-controller pattern. So, at the beginning, the application loads data from the database to ORM objects and presents them to the user. Then, the user modifies some fields or adds new objects and finally commits everything by pressing "Save" button. I experimented with this approach a bit. Problems start if ORM objects expire unintentionally, for example, due to connectivity issues or database restart, whatever (I'm not aware of all possible cases). 

the objects wouldn't expire here unless you mean that the connectivity issue is causing an error raise and then a rollback to occur.

working with detached objects wouldn't necessarily help with that if you still need lazy loading to be available; even if you merged the object temporarily just to lazy-load some collection, if you get connectivity failure + error, that object will get expired.   we dont currently have a "real" option for the Session to rollback but not expire the objects.   maybe if you did a ping on the connection directly each time a button were pushed (invalidating it on error so that it refreshses) you could prevent the Session itself from getting forced into a rollback situation.

if we are really talking about exactly **just** the ability to lazy load a relationship and nothing else, there is another option here (which I didnt recall earlier as I haven't referred to it in years).  You can use a little-known option called [enable_relationship_loading](https://docs.sqlalchemy.org/en/14/orm/session_api.html#sqlalchemy.orm.Session.enable_relationship_loading).  This was added probably for uses cases similar to this.  your object would not be attached to the session but would just be able to use its connectivity for loading relationships.   I just did a quick test and confirmed that mutated state in the detached object is unaffected when the session is rolled back.


> if I want to implement two-phase-lock mechanism for concurrent access and would like to refresh just two columns without losing the modified data in the object - seems this is also not possible. 

you can expire and/or refresh individual attributes, session.expire() and session.refresh() support that.


> From these thoughts born the idea to keep objects detached and add them to a fresh session to commit changes. But I'm not sure if this is a good approach.

I think determining what attachment behaviors you need is best here and hopefully it's just that relationship loading thing.  I agree having a single point where the objects go into "time to persist to the DB" mode is cleanest for a GUI application.

**Reply by ostapkostyk** [reply]
regarding this:
> the objects wouldn't expire here unless you mean that the connectivity issue is causing an error raise and then a rollback to occur.

and this:
> maybe if you did a ping on the connection directly each time a button were pushed (invalidating it on error so that it refreshses) you could prevent the Session itself from getting forced into a rollback situation.

Consider simple ORM class User:
```
class User(Base):
    __tablename__ = "users"
    metadata = metadata_public

    id : Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str] = mapped_column(String)
    age: Mapped[int] = mapped_column(Integer)

    def __repr__(self):
        return f"User: {self.id}, {self.name}, {self.age}"
```
I did some tests (from the console, Python 10):
```
>>> session = Session()
>>> u1 = session.query(User).order_by(User.id).first()
>>> u1
User: 1, u1, 25
>>> u1.name = 'new'
>>> u1
User: 1, new, 25
>>> session.execute(text("SELECT 1"))
<sqlalchemy.engine.cursor.CursorResult object at 0x000002C4C3B0F940>
```
now I disconnect the network and continue:
```
>>> session.execute(text("SELECT 1"))
..... I've cut part of the exception output to simplify it ....
psycopg.OperationalError: consuming input failed: server closed the connection unexpectedly
        This probably means the server terminated abnormally
        before or while processing the request.
```
now I restore connection and continue:
```
>>> session.execute(text("SELECT 1"))
sqlalchemy.exc.PendingRollbackError: Can't reconnect until invalid transaction is rolled back.  Please rollback() fully before proceeding (Background on this error at: https://sqlalche.me/e/20/8s2b)
```
so, I have to rollback the session which means that the user input will be lost. Of course, I can have separate session just to test connectivity, don't know yet how will it behave. But I tried another way:
```
>>> session.expunge(u1)
>>> session.rollback()
>>> session.execute(text("SELECT 1"))
<sqlalchemy.engine.cursor.CursorResult object at 0x000002C4C3BE6140>
>>> session.merge(u1)
User: 1, new, 25
```
then, the next step just for illustration (I tried to see what happens), can be skipped:
```
>>> session.add(u1)
sqlalchemy.exc.InvalidRequestError: Can't attach instance <User at 0x2c4c3be12a0>; another instance with key (<class 'app.models.test2.User'>, (1,), None) is already present in this session.
```
then I tried to commit a session (merged with detached object), and to read the same row (there is only one row in a table) to check if the change persists:
```
>>> session.commit()
>>> u1_1 = session.query(User).order_by(User.id).first()
>>> u1_1
User: 1, new, 25
>>> u1
User: 1, new, 25
>>> session.close()
>>> session = Session()
>>> u1_1 = session.query(User).order_by(User.id).first()
>>> u1_1
User: 1, new, 25
>>> u1 is u1_1
False
>>> u1 == u1_1
False
```
then, to restore the connection of original objects with the database I do this:
```
>>> session.close()
>>> session = Session()
>>> session.add(u1)
>>> u1
User: 1, new, 25
```
now everything is back to original state (mapped objects are again synchronized with the database)

As we can see, the change persists and modifications are safe when stored in a detached object. So, the merge can work. But I have not tried more complicated scenarios with related objects yet. If it works the same way, then the solution might be as follows:  
- application keeps the session alive all the time
- app does "ping" regularly
- if session brakes (after "ping" or after failed attempt to commit), the app detaches objects, tries to re-connect (user action required), then merge detached objects to a new session and commits. This way, user can try several attempts to save changes.
- app opens a new session and adds detached objects to it. At this point everything is restored to its original state. 

What do you think about it? I just started with SQLAlchemy one week ago and don't know it deeply enough, maybe there are some catches in the above procedure? Among others, I'm thinking about changing lazy loading to select-in to prevent expire of the object if the user just wants to access some of the attributes that were not accessed yet and at that moment it appears that the connection is broken.

EDIT:
I tried also to use a second session just to check the connection, something like:
```python
def check_connection():
    with Session() as session:
        try:
            session.execute(text("SELECT 1"))
            return True
        except:
            return False
```
to check the connection before committing changes or other operations like append. This check can be added to block some GUI buttons etc.  
Of course, this doesn't guarantee that the connection will not fail after the check and during saving user changes. So both procedures can be used: to check connectivity before accessing the database AND recovery procedure as described above.

**Reply by zzzeek** [reply]
that's not what I meant by "ping on the connection directly".  I meant this, which note, we are not going to do, because for it to work in SQLAlchemy 2.0, you need to manipulate private APIs:

```py

#### not my recommendation, this isn't really supported right now ####

try:
    s.connection().scalar(text("select 1"))
except Exception as err:
    print(f"got ping exception {e}")
    conn = s.connection()
    conn.rollback()
    trans = conn.begin()
    # 2.0 won't work without doing this, which is not public API , so this is not 
    # a real option right now
    s._transaction._connections[e] = s._transaction._connections[conn] = (
        conn,
        trans,
        True,
        True,
    )
```

I have a script where I tested the above and it works but it's not a pattern we've supported.  We'd need to add a semi-public method to Session in order to achieve the above without compatibility concerns.

I think your best option is to detach the objects and use ``enable_relationship_loading()`` to use the `Session` in a limited way prior to the "commit", then put all the objects into the `Session` at once when ready to commit.

I'd be careful about ``merge()`` because by default it queries the database for all the objects given, which is wasteful in this case.

**Reply by zzzeek** [reply]
overall I would also add that the use case described here, we really don't have any mode of operation that supports it.  many years ago when the Session was much less automated, it would support this case as it was very passive about things.   There are API features that have been removed that would support this case, namely a parameter called ``_enable_transaction_accounting=False`` that was put in place around version 0.5 when we added the new transaction accounting features, to support legacy apps that needed it turned off.   only one person needed it at that time. 

what this case would need is a Session which :

1. places the connection into DBAPI autocommit mode, holds onto it and uses it for querying
2. turns off all transaction accounting completely, snapshots etc. are not used at all
3. the flush() method would disable DBAPI autocommit mode, start a transaction and commit or rollback, re-enable DBAPI autocommit.    

The original Session worked this way with the exception that DBAPI autocommit mode didn't exist, it used its own "autocommit" mode that we have now removed in 2.0, as it did connection pool checkouts every time and introduced a whole layer of new use patterns we didnt want to support.

for the disconnects, this is really something we have never done, which would be:

4. optionally a way to have Connection invalidate and reconnect automatically without ending "the transaction", since its autocommit anyway.    this is a hacky way to do that

```py

@event.listens_for(e, "handle_error")
def handle_error(ctx):
    if ctx.is_disconnect:
        ctx.is_disconnect = False
        fairy = ctx.connection.connection
        fairy.invalidate()
        ctx.connection._dbapi_connection = fairy._pool.connect()
```


overall none of this is within the Session's design goals for a long time now and this would be all new.   For an application that is trying to maintain it's own "client side transaction" as you are doing, we always considered that to be out of scope; you'd likely use a new Session for every operation and close it each time, transferring the state from the row-based ORM objects into the view layer which is a non-ORM data structure.

**Reply by zzzeek** [reply]
OK, so we have one escape hatch, which is, `Session.close()` does not change the objects.   That's how to get around the "accounting" in most cases.    if you use a Session for every GUI operation, then close it at the end.  never call rollback.

the one place that doesn't work is if you have a disconnect error within flush(), which you do only at the point of "save".    so you have to use merge there.

here's a proof of concept of this approach:


```py
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy import text
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import sessionmaker

Base = declarative_base()


class A(Base):
    __tablename__ = "a"

    id = Column(Integer, primary_key=True)
    data = Column(String)


e = create_engine(
    "postgresql://scott:tiger@localhost/test", echo=True, echo_pool="debug"
)

Base.metadata.drop_all(e)
Base.metadata.create_all(e)

Session = sessionmaker(e, expire_on_commit=False, autoflush=False)

# GUI BUTTON PUSHED, MAKE A NEW A
with Session() as s:
    a1 = A(data="a1")
    s.add(a1)
    s.commit()

# GUI BUTTON PUSHED, LOAD AN A, CHANGE SOMETHING ON IT
with Session() as s:
    a1 = s.query(A).first()

    # make a pending change, don't commit it
    a1.data = "d2"


# GUI BUTTON PUSHED, BUT THE DB WILL DISCONNECT IN THE MIDDLE
try:
    # show an operation that needs a restart
    with Session() as s:

        # use .add() if we are not flushing
        s.add(a1)

        print("\n\n\n")
        input("restart the database")

        s.execute(text("select 1"))
except:
    print("got an error but we keep going")

# didn't lose the data
assert a1.data == "d2"


# GUI BUTTON PUSHED, SAVE EVERYTHING.  DB IS GOING TO DISCONNECT AGAIN
# this is the "persist" case, use merge.
with Session() as s:

    new_a1 = A(data='d3')
    a1_merged = s.merge(a1)
    new_a1_merged = s.merge(new_a1)

    print("\n\n\n")
    input("restart the database")

    try:
        s.commit()
    except:
        pass

# objects were not affected, can try again
assert new_a1.data == 'd3'
assert a1.data == "d2"

```

**Reply by ostapkostyk** [reply]
Thanks for your proof of concept! Yesterday I was experimenting with it. But I added more complexity to my tests - I wanted to see how will it work with related objects. It was not very straightforward but seems I grasped a concept.  

If I want to create a child object, I need to add a parent object to that session first. Or when I read child objects out of database, I need to add parent objects within that session first. Then, even if I close that session and all objects change to detached state, they hold relations with each other. So when attributes of child objects change, I just need to add/merge a parent (or root in the case of the tree structure) object to a session to commit or refresh also related objects.  

and I use ```expire_on_commit=False, autoflush=False``` flags in a session and ```.options(joinedload(MyClass.relation))``` when querying objects.  

Then, I want to have one function to save existing or new objects, so I experimented with such "save" function:
```python
def save(a):
    with Session() as session:
        try:
            session.execute(text("select 1"))

            if inspect(a).has_identity:
                session.merge(a)
            else:
                session.add(a)

            session.commit()
        except:
            print('saving failed')
```

I'm going to start with the concept to keep objects detached and with short-living sessions. Thanks a lot for your help!

**Reply by CaselIT** [reply]
If you plan on using short lived sessions you may also more simply use pool_pre_ping https://docs.sqlalchemy.org/en/14/core/engines.html#sqlalchemy.create_engine.params.pool_pre_ping

**Reply by ostapkostyk** [reply]
I still have problems when using detached objects. I'd like to use relations between objects to be able to access parents and children in a tree structure. I can do it to some extent, but get exceptions in two cases (up to now):  
- after refreshing some objects in a tree
- when trying to access a parent of an object

First case:
```python
from sqlalchemy import create_engine, Integer, String, MetaData, ForeignKey, inspect
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column, relationship, sessionmaker, joinedload

metadata = MetaData(schema="test")

engine = create_engine(f"postgresql+psycopg://{db_user}:{db_pwd}@{host}/{db_name}", echo=False, pool_pre_ping=True)
#
Session = sessionmaker(bind=engine, expire_on_commit=False, autoflush=False)

class Base(DeclarativeBase):
    pass

class A(Base):
    __tablename__ = "a"
    metadata = metadata

    id: Mapped[int] = mapped_column(primary_key=True)
    data: Mapped[str] = mapped_column(String)

    b: Mapped[list["B"]] = relationship(back_populates="a", cascade="all, delete-orphan")

    def __repr__(self):
        return f"A: {self.id}, {self.data}"


class B(Base):
    __tablename__ = "b"
    metadata = metadata

    id: Mapped[int] = mapped_column(primary_key=True)
    a_fk: Mapped[int] = mapped_column(ForeignKey("a.id"))
    data: Mapped[str] = mapped_column(String)

    a: Mapped["A"] = relationship(back_populates="b")   # parent
    c: Mapped[list["C"]] = relationship(back_populates="bc", cascade="all, delete-orphan") # children

    def __repr__(self):
        return f"B: id={self.id}, a_fk={self.a_fk}, data={self.data}"


class C(Base):
    __tablename__ = "c"
    metadata = metadata

    id: Mapped[int] = mapped_column(primary_key=True)
    b_fk: Mapped[int] = mapped_column(ForeignKey("b.id"))
    data: Mapped[str] = mapped_column(String)

    bc: Mapped["B"] = relationship(back_populates="c")

    def __repr__(self):
        return f"C: id={self.id}, b_fk={self.b_fk}, data={self.data}"


metadata.drop_all(bind=engine)
metadata.create_all(bind=engine)

a1 = A(data='a1')
b1 = B(data='b1')
a1.b.append(b1)
c1 = C(data='c1')
b1.c.append(c1)

print(c1.bc.a)    # ok

with Session() as s:
    s.add(a1)
    s.commit()

print(f"{a1}, {b1}, {c1}")   # ok
print(c1.bc.a)                  # ok

with Session() as s:
    s.add(a1)

print(a1.b)     # works, relations are still there (after the root object has been added to a session)
print(c1.bc.a)  # works also in the other direction

with Session() as s:
    s.add(a1)
    s.refresh(a1)

print(a1.b)     # relations do not work after refresh, lazy load demanded: sqlalchemy.orm.exc.DetachedInstanceError:
                # Parent instance <A at 0x...> is not bound to a Session; lazy load operation of attribute 'b' cannot
                # proceed
```

And the second case, that I can't get parent object:
```python
from sqlalchemy import create_engine, Integer, String, MetaData, ForeignKey, inspect
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column, relationship, sessionmaker, joinedload

metadata = MetaData(schema="test")

engine = create_engine(f"postgresql+psycopg://{db_user}:{db_pwd}@{host}/{db_name}", echo=False, pool_pre_ping=True)
#
Session = sessionmaker(bind=engine, expire_on_commit=False, autoflush=False)

class Base(DeclarativeBase):
    pass

class User(Base):
    __tablename__ = "users"
    metadata = metadata

    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str] = mapped_column(String)

    addresses: Mapped[list["Address"]] = relationship(back_populates="user", cascade="all, delete-orphan")

    def __repr__(self):
        return f"User: {self.id}, {self.name}"


class Address(Base):
    __tablename__ = "addresses"
    metadata = metadata

    id: Mapped[int] = mapped_column(primary_key=True)
    user_fk: Mapped[int] = mapped_column(ForeignKey("users.id"))
    address: Mapped[str] = mapped_column(String)

    user: Mapped["User"] = relationship(back_populates="addresses")

    def __repr__(self):
        return f"Address: {self.id}, {self.address}, user_ref:{self.user_fk}"


metadata.drop_all(bind=engine)
metadata.create_all(bind=engine)

user1 = User(name="user1")
user1.addresses.append(Address(address='adr1'))

with Session() as s:
    s.add(user1)
    s.commit()

addr1 = user1.addresses[0]

assert(addr1.address == 'adr1')     # attribute available without accessing database

parent = addr1.user   # this works, can access parent in detached state
assert(parent.name == 'user1')

# now load it from scratch:
with Session() as s:
    users = s.query(User).options(joinedload(User.addresses)).order_by(User.id).all()

user1 = users[0]

assert(user1.name == 'user1')

addr1 = user1.addresses[0]  # can access child members in detached state (thanks to joinedload option)
assert(addr1.address == 'adr1')

parent = addr1.user  # can't access parent in detached state, exception sqlalchemy.orm.exc.DetachedInstanceError
assert(parent.name == 'user1')
```

Is there a way to keep all relations in a detached state?

**Reply by zzzeek** [reply]
the first case where you use refresh(), you can't use refresh here because it expires attributes.  a.b is a lazyloaded so it does not get loaded, stays expired.

Use populate_existing instead so you can also load A.b

```py
    s.get(A, inspect(a1).identity, populate_existing=True, options=[selectinload(A.b)])
```

second case, make sure Address.user is also loaded

```py
with Session() as s:
    users = (
        s.query(User)
        .options(joinedload(User.addresses).immediateload(Address.user))
        .order_by(User.id)
        .all()
    )

```

**Comment by jvanasco** [comment]
FWIW, I've made GUI applications with SQLAlchemy in the past. I used the detatch+merge pattern BUT i also had either a revision-id or last-edit timestamp on each column. Edit operations were required to be against the last timestamp/revision, or would fail.  This protected against race conditions caused by multiple users or application instances, which was a concern because the application was run on Mac machines that had multiple users.

**Reply by ostapkostyk** [reply]
Thanks for sharing your experience! I was also thinking about concurrent access and wanted to implement it as two-phase locks for each row. 
Timestamp-based concurrency control might be even better.
